### PR TITLE
Add Speech sync recognize system test.

### PR DESCRIPTION
**Pull request by daspecster**
First swing at system tests for sync_recognize().

The GCS test is using a file(hello.wav) on my bucket so we'll probably want to upload that to the testing account.

### CODE REVIEW

File: system_tests/run_system_test.py [hunk 1]
```diff
@@ -24,13 +24,15 @@
 import monitoring
 import pubsub
 import storage
+import speech
```
Comment (dhermes) [inline system_tests/run_system_test.py]: alphabetical order: `speech < storage`

File: system_tests/speech.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import unittest
+
+
+class TestSpeechClient(unittest.TestCase):
+
+    def test_sync_recognize_local_file(self):
+        import io
+        from google.cloud import speech
+        client = speech.Client()
+        file_name = 'system_tests/data/hello.wav'
```
Comment (dhermes) [inline system_tests/speech.py]: This is a UNIX-path. Use `os.path.join`

File: system_tests/speech.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+        with io.open(file_name, 'rb') as file_obj:
```
Comment (dhermes) [inline system_tests/speech.py]: Why using `io.open`? Is `open(file_name, 'rb')` not sufficient?

File: system_tests/speech.py (hunk 2 above)
Comment (dhermes) [inline system_tests/speech.py]: This file isn't checked in, is it?

File: system_tests/speech.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                                        profanity_filter=True,
+                                        speech_context=['Google', 'cloud'])
+            self.assertEqual(res[0].transcript,
+                             'hello thank you for using Google Cloud platform')
+            self.assertEqual(len(res), 2)
+
+    def test_sync_recognize_gcs_file(self):
+        from google.cloud import speech
+        client = speech.Client()
+        source_uri = 'gs://ferrous-arena-my-test-bucket/hello.wav'
```
Comment (dhermes) [inline system_tests/speech.py]: What do we want to do here? I think an env. var. for the bucket name is the "right way" to go?

File: system_tests/speech.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                                        max_alternatives=2,
+                                        profanity_filter=True,
+                                        speech_context=['Google', 'cloud'])
+            self.assertEqual(res[0].transcript,
+                             'hello thank you for using Google Cloud platform')
+            self.assertEqual(len(res), 2)
+
+    def test_sync_recognize_gcs_file(self):
+        from google.cloud import speech
+        client = speech.Client()
```
Comment (dhermes) [inline system_tests/speech.py]: Can you leave an empty line between imports and code? (I realize we don't do that in all tests, but that doesn't mean we shouldn't.)

File: system_tests/speech.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+                               encoding=speech.Encoding.LINEAR16,
+                               sample_rate=16000)
+        res = client.sync_recognize(sample,
+                                    language_code='en-US',
+                                    max_alternatives=1,
+                                    profanity_filter=True,
+                                    speech_context=['Google', 'cloud'])
+        self.assertEqual(res[0].transcript,
+                         'hello thank you for using Google Cloud platform')
+        self.assertEqual(len(res), 1)
```
Comment (dhermes) [inline system_tests/speech.py]: Everything from `res = client.sync_recognize` is identical to above? Factor out a helper that takes a `sample` and does all the verification.

File: system_tests/speech.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        import io
+        from google.cloud import speech
+        client = speech.Client()
+        file_name = 'system_tests/data/hello.wav'
+
+        with io.open(file_name, 'rb') as file_obj:
+            sample = client.sample(content=file_obj.read(),
+                                   encoding=speech.Encoding.LINEAR16,
+                                   sample_rate=16000)
+            res = client.sync_recognize(sample,
```
Comment (dhermes) [inline system_tests/speech.py]: Instead of `res`, `result`?

File: system_tests/speech.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+                                        language_code='en-US',
+                                        max_alternatives=2,
+                                        profanity_filter=True,
+                                        speech_context=['Google', 'cloud'])
+            self.assertEqual(res[0].transcript,
```
Comment (dhermes) [inline system_tests/speech.py]: Please verify properties other than `transcript` (i.e. all of them)

File: system_tests/speech.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+                             'hello thank you for using Google Cloud platform')
+            self.assertEqual(len(res), 2)
```
Comment (dhermes) [inline system_tests/speech.py]: If there are **two** results we should verify both of them

File: system_tests/speech.py (hunk 2 above)
Comment (daspecster) [inline system_tests/speech.py]: It's new in this PR, https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2547/files#diff-cfdb0984855e984019c4825998562444.

Comment (dhermes) [inline system_tests/speech.py]: D'oh!

File: system_tests/speech.py (hunk 3 above)
Comment (daspecster) [inline system_tests/speech.py]: So I had to look into this a lot, but in the end `open()` is an alias to `io.open()`.  Which I guess isn't true in Python2.6?

I wanted the `BufferedReader` for binary reading of the file and I thought builtin `open()` didn't always use that when you expected it would.

However if you run `help(io.open)` it says...
> Help on built-in function open in module _io...

So I guess it is just an alias?

Comment (dhermes) [inline system_tests/speech.py]: The `b` in `rb` makes it binary. https://docs.python.org/2/tutorial/inputoutput.html#reading-and-writing-files

File: system_tests/speech.py (hunk 4 above)
Comment (daspecster) [inline system_tests/speech.py]: Sure, is there one that's being used already that I'm missing?

Comment (daspecster) [inline system_tests/speech.py]: I'm going with `SPEECH_GCS_URI` for now.

File: system_tests/run_system_test.py (hunk 1 above)
Comment (daspecster) [inline system_tests/run_system_test.py]: Man....I guess I need to go watch sesame street again.

**Comment by daspecster** [comment]
@dhermes, I had to change how the transcript and confidence values are accessed. I think I was running this before with a different version installed so I had them as properties. The result for `sync _recognize` is currently a dictionary.

**Comment by dhermes** [comment]
> I had to change how the transcript and confidence values are accessed. I think I was running this before with a different version installed so I had them as properties

It's tough to review a moving target. Are the tested methods done and checked in? If yes, then why the out-of-sync issue?

File: system_tests/speech.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        return result
+
+    def test_sync_recognize_local_file(self):
+        file_name = 'system_tests/data/hello.wav'
+
+        with open(file_name, 'rb') as file_obj:
+            result = self._make_sync_request(content=file_obj.read(),
+                                             max_alternatives=2)
+
+            self.assertEqual(result[0]['transcript'],
```
Comment (dhermes) [inline system_tests/speech.py]: Big time :-1: on `result[0]` being a dictionary. Didn't you create custom types for this?

PS I realize this PR is not the place to **change** the API surface, though we should discuss while it's in view.

File: system_tests/speech.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+                             'hello thank you for using Google Cloud platform')
+            self.assertGreater(result[0]['confidence'], .90)
+            self.assertEqual(result[1]['transcript'],
+                             'thank you for using Google Cloud platform')
+            self.assertEqual(len(result), 2)
+
+    def test_sync_recognize_gcs_file(self):
+        import os
+
+        source_uri = os.getenv('SPEECH_GCS_URI')
```
Comment (dhermes) [inline system_tests/speech.py]: Rather than using an env. var., do what the `language` system tests do: [upload][2] to a [test local bucket][1] and then [read that object][3] (and [clean up after][4]).

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/881a95330dfb0dbab71e9c405acd2a8423ff1fee/system_tests/language.py#L35-L43
[2]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/881a95330dfb0dbab71e9c405acd2a8423ff1fee/system_tests/language.py#L106-L110
[3]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/881a95330dfb0dbab71e9c405acd2a8423ff1fee/system_tests/language.py#L114
[4]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/881a95330dfb0dbab71e9c405acd2a8423ff1fee/system_tests/language.py#L46-L51

File: system_tests/speech.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        result = client.sync_recognize(sample,
+                                       language_code='en-US',
+                                       max_alternatives=max_alternatives,
+                                       profanity_filter=True,
+                                       speech_context=['Google',
+                                                       'cloud'])
+        return result
+
+    def test_sync_recognize_local_file(self):
+        file_name = 'system_tests/data/hello.wav'
```
Comment (dhermes) [inline system_tests/speech.py]: [This is a UNIX-path. Use `os.path.join`][1]

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2547#discussion_r83706763

File: system_tests/speech.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+        with open(file_name, 'rb') as file_obj:
+            result = self._make_sync_request(content=file_obj.read(),
+                                             max_alternatives=2)
+
+            self.assertEqual(result[0]['transcript'],
+                             'hello thank you for using Google Cloud platform')
+            self.assertGreater(result[0]['confidence'], .90)
```
Comment (dhermes) [inline system_tests/speech.py]: Please also verify that `confidence` and `transcript` are the only keys in `result[0]` (similarly for `result[1]`, assuming `confidence` **is not** a key in thay one)

File: system_tests/speech.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+# limitations under the License.
+
+import unittest
+
+
+class TestSpeechClient(unittest.TestCase):
+
+    def _make_sync_request(self, content=None, source_uri=None,
+                           max_alternatives=None):
+        from google.cloud.speech.encoding import Encoding
```
Comment (dhermes) [inline system_tests/speech.py]: `Encoding` hasn't been put in `speech.__init__` yet?

File: system_tests/speech.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+from google.cloud import speech
+from google.cloud import storage
+
+from system_test_utils import unique_resource_id
+from retry import RetryErrors
+
+
+class Config(object):
+    """Run-time configuration to be modified at set-up.
+    This is a mutable stand-in to allow test set-up to modify
```
Comment (dhermes) [inline system_tests/speech.py]: Empty line between the one-liner and the rest

File: system_tests/speech.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+        sample = client.sample(content=content,
+                               source_uri=source_uri,
+                               encoding=speech.Encoding.LINEAR16,
+                               sample_rate=16000)
+        result = client.sync_recognize(sample,
+                                       language_code='en-US',
+                                       max_alternatives=max_alternatives,
+                                       profanity_filter=True,
+                                       speech_context=['Google',
+                                                       'cloud'])
```
Comment (dhermes) [inline system_tests/speech.py]: Can fit on the same line?

File: system_tests/speech.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+        return result
+
+    def _check_best_results(self, results):
+        from google.cloud.speech.transcript import Transcript
```
Comment (dhermes) [inline system_tests/speech.py]: Since you are doing imports globally, why is this here?

File: system_tests/speech.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+
+        top_result = results[0]
+        self.assertIsInstance(top_result, Transcript)
+        self.assertEqual(top_result.transcript,
+                         'hello thank you for using Google Cloud platform')
+        self.assertGreater(top_result.confidence, 0.90)
+
+    def test_sync_recognize_local_file(self):
+        import os
+        from google.cloud.speech.transcript import Transcript
```
Comment (dhermes) [inline system_tests/speech.py]: Ditto on imports

File: system_tests/speech.py [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+
+        file_name = os.path.join('system_tests', 'data', 'hello.wav')
```
Comment (dhermes) [inline system_tests/speech.py]: This is a relative path and will depend on where the file gets run from. You should use `os.path.dirname(__file__)` instead of `system_tests`.

It's almost certainly worth making the value of `os.path.dirname(__file__)` a global and then referring to it when you make paths. But it's also fine to make the actual path to `hello.wav` be a global.

File: system_tests/speech.py [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+        with open(file_name, 'rb') as file_obj:
+            results = self._make_sync_request(content=file_obj.read(),
+                                              max_alternatives=2)
+            second_alternative = results[1]
+            self.assertEqual(len(results), 2)
+            self._check_best_results(results)
+            self.assertIsInstance(second_alternative, Transcript)
+            self.assertEqual(second_alternative.transcript,
+                             'thank you for using Google Cloud platform')
+            self.assertEqual(second_alternative.confidence, 0.0)
```
Comment (dhermes) [inline system_tests/speech.py]: `0.0` is the null value in the proto. Shouldn't `second_alternative.confidence` be `None`? Is it also `0.0` when you use the JSON API?

File: system_tests/speech.py [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+
+        with open(file_name, 'rb') as file_obj:
+            results = self._make_sync_request(content=file_obj.read(),
+                                              max_alternatives=2)
+            second_alternative = results[1]
+            self.assertEqual(len(results), 2)
+            self._check_best_results(results)
+            self.assertIsInstance(second_alternative, Transcript)
+            self.assertEqual(second_alternative.transcript,
+                             'thank you for using Google Cloud platform')
```
Comment (dhermes) [inline system_tests/speech.py]: Maybe `'thank you for using Google Cloud platform'` should be a class constant since you use it here and elsewhere. The you could say

```python
self.assertEqual(second_alternative.transcript, self.FOO)

# AND above

self.assertEqual(top_result.transcript, 'hello ' + self.FOO)
```

File: system_tests/speech.py [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+
+        file_name = os.path.join('system_tests', 'data', 'hello.wav')
+        bucket_name = Config.TEST_BUCKET.name
+        blob_name = 'document.txt'
+        blob = Config.TEST_BUCKET.blob(blob_name)
+        self.to_delete_by_case.append(blob)  # Clean-up.
+        with open(file_name, 'rb') as file_obj:
+            blob.upload_from_file(file_obj)
+
+        # source_uri = os.getenv('SPEECH_GCS_URI')
```
Comment (dhermes) [inline system_tests/speech.py]: Please remove commented out code.

File: system_tests/speech.py [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+            self._check_best_results(results)
+            self.assertIsInstance(second_alternative, Transcript)
+            self.assertEqual(second_alternative.transcript,
+                             'thank you for using Google Cloud platform')
+            self.assertEqual(second_alternative.confidence, 0.0)
+
+    def test_sync_recognize_gcs_file(self):
+        import os
+
+        file_name = os.path.join('system_tests', 'data', 'hello.wav')
```
Comment (dhermes) [inline system_tests/speech.py]: Ditto here on the path. Just make it a global since you use it twice.

File: system_tests/speech.py [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+        bucket_name = Config.TEST_BUCKET.name
+        blob_name = 'document.txt'
```
Comment (dhermes) [inline system_tests/speech.py]: Why wouldn't you name the blob `hello.wav`?

File: system_tests/speech.py (hunk 20 above)
Comment (daspecster) [inline system_tests/speech.py]: It's always an instance of the `Transcript` class so it's the same for JSON/gRPC.

I can change it to None but it'll be a separate PR. Before or after this one though?

File: system_tests/speech.py (hunk 24 above)
Comment (daspecster) [inline system_tests/speech.py]: Gah oops.

File: system_tests/speech.py [hunk 25]
```diff
@@ -12,23 +12,28 @@
...
 class Config(object):
     """Run-time configuration to be modified at set-up.
+
     This is a mutable stand-in to allow test set-up to modify
     global state.
     """
     CLIENT = None
     TEST_BUCKET = None
+    AUDIO_FILE = os.path.join(os.path.dirname(__file__), 'data', 'hello.wav')
+    ASSERT_TEXT = 'thank you for using Google Cloud platform'
```
Comment (dhermes) [inline system_tests/speech.py]: @daspecster Define this on `TestSpeechClient`

File: system_tests/speech.py [hunk 26]
```diff
@@ -12,23 +12,28 @@
...
 
 class Config(object):
     """Run-time configuration to be modified at set-up.
+
     This is a mutable stand-in to allow test set-up to modify
     global state.
     """
     CLIENT = None
     TEST_BUCKET = None
+    AUDIO_FILE = os.path.join(os.path.dirname(__file__), 'data', 'hello.wav')
```
Comment (dhermes) [inline system_tests/speech.py]: Make this a global. This isn't "Run-time configuration", i.e. it's always that value at import time, doesn't differ between the actual runs of this system test.

File: system_tests/speech.py [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+    def test_sync_recognize_local_file(self):
+        with open(Config.AUDIO_FILE, 'rb') as file_obj:
+            results = self._make_sync_request(content=file_obj.read(),
+                                              max_alternatives=2)
+            second_alternative = results[1]
+            self.assertEqual(len(results), 2)
+            self._check_best_results(results)
+            self.assertIsInstance(second_alternative, Transcript)
+            self.assertEqual(second_alternative.transcript, Config.ASSERT_TEXT)
+            self.assertEqual(second_alternative.confidence, 0.0)
```
Comment (dhermes) [inline system_tests/speech.py]: RE:

> It's always an instance of the `Transcript` class so it's the same for JSON/gRPC.
>
> I can change it to None but it'll be a separate PR. Before or after this one though?

Though it is always an instance of `Transcript` (I don't think that's a PB/JSON type), the JSON default may just be to leave out the value, while the protobuf treats `0.0` like it's empty:

```python
>>> from google.cloud.grpc.speech.v1beta1.cloud_speech_pb2 import (
...     SpeechRecognitionAlternative)
>>> transcript = SpeechRecognitionAlternative()
>>> transcript.confidence
0.0
>>> print(transcript)

>>> transcript.transcript = 'hi mom'
>>> print(transcript)
transcript: "hi mom"

>>> transcript.confidence = 1.0
>>> print(transcript)
transcript: "hi mom"
confidence: 1.0

>>> transcript.confidence = 0.0
>>> print(transcript)
transcript: "hi mom"

```

Comment (daspecster) [inline system_tests/speech.py]: >[Output-only] The confidence estimate between 0.0 and 1.0. A higher number means the system is more confident that the recognition is correct. This field is typically provided only for the top hypothesis, and only for isFinal=true results. **The default of 0.0 is a sentinel value indicating confidence was not set.**



Ref: https://cloud.google.com/speech/reference/rest/v1beta1/speech/syncrecognize#SpeechRecognitionAlternative

Comment (dhermes) [inline system_tests/speech.py]: Right, that's what I'm saying.

Comment (daspecster) [inline system_tests/speech.py]: That's the docs for the REST API though.

Comment (dhermes) [inline system_tests/speech.py]: Ah I see. Yeah, punt on this until another PR, but do as the docs say: turn `0.0` into `None`.
